src/IK.cpp

```diff
@@ -0,0 +1,103 @@
+// This file is part of RBDyn.
+//
+// RBDyn is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// RBDyn is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with RBDyn.  If not, see <http://www.gnu.org/licenses/>.
+
+// associated header
+#include "IK.h"
+
+// includes
+// RBDyn
+#include "FK.h"
+#include "FV.h"
+#include "MultiBody.h"
+#include "MultiBodyConfig.h"
+
+//SpaceVecAlg
+#include <SpaceVecAlg/SpaceVecAlg>
+
+namespace rbd
+{
+
+static const int MAX_IK_ITERATIONS = 50;
+static const double LAMBDA = 0.9;
+static const double IK_THRESHOLD = 1e-8;
+static const double ALMOST_ZERO = 1e-8;
+
+InverseKinematics::InverseKinematics(const MultiBody& mb, int ef_index):
+	ef_index_(ef_index),
+	jac_(mb, mb.body(ef_index).id()),
+	svd_()
+{
+}
+
+struct CwiseRoundOp {
+	CwiseRoundOp(const double& inf, const double& sup) : m_inf(inf), m_sup(sup) {}
+	double operator()(const double& x) const { return x>m_inf && x<m_sup ? 0 : x; }
+	double m_inf, m_sup;
+};
+
+bool InverseKinematics::inverseKinematics(const MultiBody& mb, MultiBodyConfig& mbc,
+                                          const sva::PTransformd& ef_target)
+{
+	int iter = 0;
+	bool converged = false;
+	int dof = 0;
+	rbd::forwardKinematics(mb, mbc);
+	Eigen::MatrixXd jacMat;
+	Eigen::Vector6d v = Eigen::Vector6d::Ones();
+	Eigen::Vector3d rotErr;
+	Eigen::VectorXd res = Eigen::VectorXd::Zero(3);
+	while(iter < MAX_IK_ITERATIONS && ! converged)
```

Comment:
Very minor but
`while( ! converged && iter < MAX_IK_ITERATIONS)`
would read better imho (and perform one less check after convergence but that's very very minor)